servers/republik/modules/crowdfundings/lib/Pledge/cancel.jest.js

```diff
@@ -0,0 +1,535 @@
+        },
+        {
+          id: 'period-3',
+          membershipId: 'membership-a',
+          beginDate: moment().add(20, 'days'),
+          endDate: moment().add(30, 'days'),
+          pledgeId: 'pledge-C'
+        }
+      ]
+
+      expect(UUT.evaluatePeriods({ pledgeId, membership, periods }, { now }))
+        .toMatchObject([
+          {
+            _raw: periods[0],
+            isCausedByPledge: false,
+            isObsolete: false,
+            updateAttributes: {}
+          },
+          {
+            _raw: periods[1],
+            isCausedByPledge: true,
+            isObsolete: true,
+            updateAttributes: {}
+          },
+          {
+            _raw: periods[2],
+            isCausedByPledge: false,
+            isObsolete: false,
+            updateAttributes: {
+              beginDate: now,
+              endDate: now.add(getInterval(periods[2].beginDate, periods[2].endDate))
+            }
+          }
+        ])
+    })
+
+    test('Pledge B caused 2 middle, future periods: flag periods obsolete and update subsequent period', () => {
+      const now = moment()
+      const pledgeId = 'pledge-B'
+      const membership = {
+        id: 'membership-a',
+        pledgeId: 'pledge-A'
+      }
+
+      const periods = [
+        {
+          id: 'period-1',
+          membershipId: 'membership-a',
+          beginDate: moment().subtract(10, 'days'),
+          endDate: moment().add(10, 'days')
+        },
+        {
+          id: 'period-2',
+          membershipId: 'membership-a',
+          beginDate: moment().add(10, 'days'),
+          endDate: moment().add(20, 'days'),
+          pledgeId: 'pledge-B'
+        },
+        {
+          id: 'period-3',
+          membershipId: 'membership-a',
+          beginDate: moment().add(20, 'days'),
+          endDate: moment().add(22, 'days'),
+          pledgeId: 'pledge-B'
+        },
+        {
+          id: 'period-4',
+          membershipId: 'membership-a',
+          beginDate: moment().add(22, 'days'),
+          endDate: moment().add(32, 'days'),
+          pledgeId: 'pledge-C'
+        }
+      ]
+
+      expect(UUT.evaluatePeriods({ pledgeId, membership, periods }, { now }))
+        .toMatchObject([
+          {
+            _raw: periods[0],
+            isCausedByPledge: false,
+            isObsolete: false,
+            updateAttributes: {}
+          },
+          {
+            _raw: periods[1],
+            isCausedByPledge: true,
+            isObsolete: true,
+            updateAttributes: {}
+          },
+          {
+            _raw: periods[2],
+            isCausedByPledge: true,
+            isObsolete: true,
+            updateAttributes: {}
+          },
+          {
+            _raw: periods[3],
+            isCausedByPledge: false,
+            isObsolete: false,
+            updateAttributes: {
+              beginDate: now,
+              endDate: now.add(getInterval(periods[3].beginDate, periods[3].endDate))
```

Comment:
should the `beginDate` here not be `periods[0].endDate`?

Reply:
True! You found a bug.

Reply:
[Adopted](6aae469) and [added and fixed more tests](ce321d6) to ensure its workings.

In theory there is a matrix of tests in 3 dimensions:
1. Point in time when cancelling happens
2. Pledges causing periods (1 period, 2 periods, etc.)
3. Pledge to cancel

Which results in about 50 to 74 scenarios aka tests.

I am no wondering if it's worth covering all of them? What do you think?

Reply:
I think the coverage is enough for now, let's add tests in case we discover a bug down the road.